### PR TITLE
feat: Add Auth0/audience support for ClientCredentials flow

### PR DESCRIPTION
**Pull request by PudgyPigeon**
# TL;DR
This PR enables passing in the `audience` parameter into the `get_token` method - via the ClientCredentials flow and the ClientConfig object. This enables using Auth0 as an external server for intra-service authentication. Thus, every aspect of auth within a Flyte deployment can be offloaded to Auth0 instead of being handled internally.

## Type
 - [x] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_
 
 Auth0 requires an audience parameter for its various requests. This was previously not enabled in the Flytekit source code as the various dataclasses and objects never expected an `audience` param as an input arg. Furthermore, the format necessary for an Auth0 request was not defined. This PR adds support for that by piping in the `audience` parameter as input from FlyteAdmin's audience field, and then into the various functions/methods/classes that handle the auth request.
 
## Tracking Issue
https://github.com/flyteorg/flyte/issues/3661

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by PudgyPigeon** [comment]
@wild-endeavor

File: flytekit/clients/auth/token_client.py [hunk 1]
```diff
@@ -94,12 +96,17 @@ def get_token(
     }
     if client_id:
         body["client_id"] = client_id
+    if client_secret:
```
Comment (wild-endeavor) [inline flytekit/clients/auth/token_client.py:99]: client secret shouldn't be sent here right?

**Comment by PudgyPigeon** [comment]
@wild-endeavor Made the changes, hopefully it is correct

File: flytekit/clients/auth/token_client.py [hunk 2]
```diff
@@ -98,8 +99,11 @@ def get_token(
         body["device_code"] = device_code
     if scopes is not None:
         body["scope"] = ",".join(scopes)
+    if audience is not None:
```
Comment (wild-endeavor) [inline flytekit/clients/auth/token_client.py:102]:
```suggestion
    if audience:
```

File: flytekit/clients/auth_helper.py [hunk 3]
```diff
@@ -66,13 +67,23 @@ def get_authenticator(cfg: PlatformConfig, cfg_store: ClientConfigStore) -> Auth
...
         return PKCEAuthenticator(cfg.endpoint, cfg_store, verify=verify)
     elif cfg_auth == AuthType.BASIC or cfg_auth == AuthType.CLIENT_CREDENTIALS or cfg_auth == AuthType.CLIENTSECRET:
-        return ClientCredentialsAuthenticator(
-            endpoint=cfg.endpoint,
-            client_id=cfg.client_id,
-            client_secret=cfg.client_credentials_secret,
-            cfg_store=cfg_store,
-            scopes=cfg.scopes,
-        )
+        if cfg.audience is not None:
```
Comment (wild-endeavor) [inline flytekit/clients/auth_helper.py:70]: can we collapse this call as well please?

**Comment by PudgyPigeon** [comment]
@wild-endeavor

**Review by wild-endeavor — COMMENTED** [review]
this looks good, but now there's merge conflicts.  would you mind addressing?  you can also give me write perms to your fork if you want and i can address.

thank you.

**Comment by PudgyPigeon** [comment]
> this looks good, but now there's merge conflicts. would you mind addressing? you can also give me write perms to your fork if you want and i can address.
> 
> thank you.

Yup, just added you as a collaborator on the fork repo

**Comment by PudgyPigeon** [comment]
@wild-endeavor latest changes from master have been merged

**Comment by PudgyPigeon** [comment]
Fixed the linting - workflows should all pass now

**Review by wild-endeavor — APPROVED** [review]
